File: 백준/Silver/2579. 계단 오르기/계단 오르기.py
import sys
n = int(sys.stdin.readline())

dp = [0] * (n + 1)
arr = [0] * (n + 1)

for i in range(1, n + 1):
    arr[i] = int(input())

if (n < 3):
    print(sum(arr[0:n+1]))
    exit()

dp[1] = arr[1]
dp[2] = arr[1] + arr[2]

for i in range(3, n + 1):
    dp[i] = max(dp[i - 2] + arr[i], dp[i - 3] + arr[i - 1] + arr[i])

print(dp[n])

File: 백준/Silver/2579. 계단 오르기/계단 오르기.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lines = [];

rl.on("line", (input) => {
  lines.push(parseInt(input))

  if (lines.length > lines[0]) {
    rl.close()
  }
});

// output
rl.on("close", () => {
  let n = lines[0]
  let arr = lines.slice(1)

  if (n === 1) {
    console.log(arr[0])
    return
  }

  let dp = new Array(n).fill(0)
  dp[0] = arr[0]
  dp[1] = arr[0] + arr[1]
  
  if (n > 2) {
    dp[2] = Math.max(arr[0] + arr[2], arr[1] + arr[2])
  }

  for (let i = 3; i < n; i++) {
    dp[i] = Math.max(dp[i - 2] + arr[i], dp[i - 3] + arr[i - 1] + arr[i])
  }
  console.log(dp[n - 1])
});
